torch/csrc/jit/codegen/cuda/lower_insert_syncs.cpp

```diff
@@ -351,31 +451,31 @@ class ReadAfterWriteSyncs : public kir::MutableIrVisitor {
   }
 
   // Clear the modify status for all shared memory buffers
-  static void cleanSharedMemory(std::unordered_map<kir::Val*, bool>& smem) {
-    for (auto& item : smem) {
-      item.second = false;
-    }
+  static void cleanSharedMemory(
+      std::unordered_map<kir::Val*, kir::Expr*>& smem) {
+    smem.clear();
   }
 
   // Return the status of the shared memory buffer
   // False if TensorView is not shared memory buffer
-  bool isModifiedSharedMemory(
-      const std::unordered_map<kir::Val*, bool>& smem,
+  std::unordered_set<kir::Expr*> isModifiedSharedMemory(
+      const std::unordered_map<kir::Val*, kir::Expr*>& smem,
       const std::vector<kir::Val*>& keys) const {
-    return std::any_of(keys.begin(), keys.end(), [&smem](kir::Val* key) {
+    std::unordered_set<kir::Expr*> last_writes;
+    for (auto key : keys) {
       auto it = smem.find(key);
       if (it != smem.end()) {
-        return it->second;
+        last_writes.insert(it->second);
       }
-      return false;
-    });
+    }
+    return last_writes;
   }
 
   ReadAfterWriteSyncs(std::vector<kir::Expr*> _loop_nests)
       : loop_nests_(std::move(_loop_nests)) {
     // Fusion shared_memory values
     // Tracks if shared memory is modified
```

Comment:
The change of `bool` to `Expr*` doesn't really affect the original logic since `bool` was only used to indicate if there's any mapping from `Val*`. `smem` didn't need to be a map but could be just a set.

What I changed is to keep track of write expressions as well. Since the boolean value wasn't necessary, I just changed the map to hold `Expr*`.

Reply:
Thank you, this makes more sense to me now.